### ISSUE TITLE
Bug: Error: Access Denied when using Gamepad Input Injection

### ISSUE THREAD

**Issue by kitswas** [microsoft/windows-rs#2574]
### Which crate is this about?

windows

### Crate version

"0.48"

### Summary

The exact error message is:
```
Error { code: HRESULT(0x80070005), message: "Access is denied." }
```

Stacktrace follows:
```
stack backtrace:
   0: std::panicking::begin_panic_handler
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library\std\src\panicking.rs:578
   1: core::panicking::panic_fmt
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library\core\src\panicking.rs:67
   2: core::result::unwrap_failed
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca/library\core\src\result.rs:1687
   3: enum2$<core::result::Result<tuple$<>,windows::core::error::Error> >::unwrap<tuple$<>,windows::core::error::Error>
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca\library\core\src\result.rs:1089
   4: pc_server_rs::input_injector::GamepadInjector::inject
             at .\src\input_injector.rs:47
   5: pc_server_rs::main
             at .\src\main.rs:21
   6: core::ops::function::FnOnce::call_once<void (*)(),tuple$<> >
             at /rustc/90c541806f23a127002de5b4038be731ba1458ca\library\core\src\ops\function.rs:250
```

### Toolchain version/configuration

Default host: x86_64-pc-windows-msvc
rustup home:  C:\Users\swast\.rustup

stable-x86_64-pc-windows-msvc (default)
rustc 1.70.0 (90c541806 2023-05-31)

### Reproducible example

```rust
//! Input injection using WinRT.
//! See https://github.com/microsoft/windows-rs
//! and https://learn.microsoft.com/en-us/uwp/api/windows.gaming.input.gamepad

use windows::Gaming::Input::GamepadButtons;
use windows::UI::Input::Preview::Injection::InjectedInputGamepadInfo;
use windows::UI::Input::Preview::Injection::InputInjector;

pub(crate) struct GamepadInjector {
    gamepad_state: InjectedInputGamepadInfo,
    injector: InputInjector,
}

impl GamepadInjector {
    pub(crate) fn new() -> Self {
        let injector: InputInjector = InputInjector::TryCreate().unwrap();
        println!("Created input injector.");
        injector.InitializeGamepadInjection().unwrap();
        println!("Initialized gamepad injection.");
        let gamepad_state = InjectedInputGamepadInfo::new().unwrap();
        Self {
            gamepad_state,
            injector,
        }
    }

    pub(crate) fn update(
        &mut self,
        buttons: GamepadButtons,
        left_thumbstick: (f64, f64),
        right_thumbstick: (f64, f64),
    ) {
        let _ = self.gamepad_state.SetButtons(buttons);
        let _ = self.gamepad_state.SetLeftThumbstickX(left_thumbstick.0);
        let _ = self.gamepad_state.SetLeftThumbstickY(left_thumbstick.1);
        let _ = self.gamepad_state.SetRightThumbstickX(right_thumbstick.0);
        let _ = self.gamepad_state.SetRightThumbstickY(right_thumbstick.1);
    }

    pub(crate) fn inject(&mut self) {
        self.injector
            .InjectGamepadInput(&self.gamepad_state)
            .unwrap();
    }
}

impl Drop for GamepadInjector {
    fn drop(&mut self) {
        self.injector.UninitializeGamepadInjection().unwrap();
    }
}
```

### Crate manifest

```toml
#...

[dependencies.windows]
version = "0.48"
features = [
	"Gaming_Input",
	"UI_Input_Preview_Injection",
]
```

### Actual behavior

Calls to `InputInjector.InjectGamepadInput()` and `InputInjector.InitializeGamepadInjection()` return the aforementioned error.

**Comment by riverar** [comment]
This API must be used in a packaged app context with the `inputInjectionBrokered` restricted capability, [per docs](https://learn.microsoft.com/en-us/uwp/api/windows.ui.input.preview.injection.inputinjector.injectgamepadinput?view=winrt-22621):

> Important
> The APIs in this namespace require the inputInjectionBrokered [restricted capability](https://learn.microsoft.com/en-us/windows/uwp/packaging/app-capability-declarations#restricted-capabilities).
> [...]

See also:
* https://learn.microsoft.com/windows/apps/design/input/input-injection
* https://learn.microsoft.com/windows/apps/desktop/modernize/desktop-to-uwp-supported-api
* https://learn.microsoft.com/windows/apps/desktop/modernize/modernize-packaged-apps

For API-specific questions, we suggest asking on:
* https://docs.microsoft.com/en-us/answers/topics/windows-api.html
* https://stackoverflow.com/questions/tagged/windows-rs

**Comment by kitswas** [comment]
Thanks for your quick response.  
I noticed the '`inputInjectionBrokered` restricted capability' requirement earlier but the examples are all in _packaging_ for C# apps. I didn't understand.

Could you mention in simple terms what it means for a Rust program?  
Is it possible to get the feature working in a Rust or a C++ app? If yes, how?

**Comment by robmikh** [comment]
You may not need to package your application. I remember calling these APIs while as admin in an unpackaged app and it working.

**Comment by riverar** [comment]
Hey @kitswas, sure. Packaging an application in this case means you take your app artifacts (exe, lib, graphics, etc.), document them via a special [manifest](https://learn.microsoft.com/uwp/schemas/appxpackage/appx-package-manifest), then box everything up with [MSIX](https://learn.microsoft.com/windows/msix/overview) (previously known as APPX). The main workflow above results in a packaged app that is put [into a container](https://learn.microsoft.com/windows/msix/msix-container), where capabilities such as injecting input and touching the registry are strictly regulated. (Think of a sandbox, if you're familiar with that term.)

We have a sample that shows an application that utilizes a manifest here https://github.com/microsoft/windows-rs/tree/master/crates/samples/windows/core_app.

But maybe none of this is needed as @robmikh points out. Does elevating your app (run as admin) solve your problem?

**Comment by kitswas** [comment]
Run as admin was the first thing I tried when I encountered the error.

On the command line,
```
powershell start target\debug\pc-server-rs.exe -Verb runas
```
I also did right-click > Run as admin from Windows Explorer.

Didn't work.  
Am I missing something?

**Comment by robmikh** [comment]
Ah interesting, it appears running in an unpackaged app without a manifest only works if you have Developer Mode enabled in the settings.

**Comment by kitswas** [comment]
Yes, the app continued to work even after I unregistered the Manifest.
That's when I noticed.
Earlier, I had enabled Developer Mode to register the Manifest.

Still couldn't get the Gamepad input injection to work, though.
The line
```
injector.InitializeGamepadInjection().unwrap();
```
doesn't work as expected (seems to do nothing).  
From [the docs](https://learn.microsoft.com/en-us/uwp/api/windows.ui.input.preview.injection.inputinjector.initializegamepadinjection?view=winrt-22621#remarks),  

> Calling this method is analogous to connecting a physical gamepad, which also triggers a [GamepadAdded](https://learn.microsoft.com/en-us/uwp/api/windows.gaming.input.gamepad.GamepadAdded) event.

I was also getting the error `thread 'main' has overflowed its stack` sporadically.

Now, I am rewriting things the Windows (official/recommended) way, i.e., using C# and following the tutorials. 

A big **thank you** to all who tried to help me.
